### PR TITLE
Feature/3902 rabbitmq mutual tls

### LINKED ISSUE #3902: [Feature] RabbitMQ Mutual TLS
Our Rabbit server is set up to require a client certificate but there appears not to be an interface in Brighter for providing one.

Some interface to provide it in the RmqMessagingGatewayConnection object would be nice.

I'll be working on an implementation for our case next week, I'm hoping someone on your side gets there before I do :)

Thank you.

### PR DESCRIPTION
**Pull request by darrenschwarz**
# Add mutual TLS (mTLS) support to RabbitMQ messaging gateways

Fixes #3902

## Summary

Adds mutual TLS (mTLS) authentication support for RabbitMQ connections in both the Sync (RMQ v6) and Async (RMQ v7) messaging gateways.

### Core Implementation

Added client certificate configuration to `RmqMessagingGatewayConnection`:
- `ClientCertificate` - X509Certificate2 object (for programmatic configuration)
- `ClientCertificatePath` - File path to .pfx/PKCS#12 certificate (for file-based configuration)
- `ClientCertificatePassword` - Optional password for certificate files
- `TrustServerSelfSignedCertificate` - Opt-in flag for test/dev environments (follows SQL Server naming convention)

**New TLS Configuration Helpers:**
- Added `RmqTlsConfigurator` classes (Async and Sync variants) to centralize TLS/SSL configuration logic
- Extracts certificate loading and `SslOption` configuration into reusable internal static classes
- Supports both X509Certificate2 objects and file paths with optional passwords
- Includes conditional compilation to support both .NET 8.0 (`X509Certificate2`) and .NET 9.0+ (`X509CertificateLoader`)
- Integrated with `RmqMessageGateway` via `RmqTlsConfigurator.ConfigureIfEnabled()` call

Fixed `GetSanitizedUri()` to handle mTLS URIs that don't contain username/password credentials (adds early return for URIs without @ symbol; existing username/password sanitization logic unchanged and backward compatible)

### Cross-Gateway Parity

Both Sync and Async gateways have identical mTLS implementations, maintaining the established pattern of feature parity between RMQ v6 and v7 clients.

### Pre-existing RabbitMQ Bugs (Discovered During Testing)

While testing the mTLS implementation, CI revealed pre-existing bugs in the RabbitMQ message requeuing logic:

**1. Missing NOT operator in AddUserDefinedHeaders (SYNC version)**
- **Issue**: The SYNC version of `AddUserDefinedHeaders` was missing the `!` negation operator that exists in the ASYNC version
- **Impact**: Only system headers were being copied during requeue instead of user-defined headers, causing message body loss
- **Fix**: Changed `if (_headersToReset.Contains(header.Key))` to `if (!_headersToReset.Contains(header.Key))`
- **Commit**: 15acf149c

These fixes bring the SYNC implementation in line with the ASYNC version.

### mTLS Implementation Bugs (Fixed in This PR)

**1. Certificate Validation Hostname Mismatch**
- **Issue**: Server certificate is issued for hostname "rabbitmq-mtls" but tests connect to "localhost", causing SSL name mismatch errors
- **Impact**: All 18 Docker-dependent mTLS tests failed with `AuthenticationException: The remote certificate was rejected by the provided RemoteCertificateValidationCallback`
- **Fix**: Updated `RmqTlsConfigurator` (both Async and Sync) to accept both `RemoteCertificateChainErrors` AND `RemoteCertificateNameMismatch` when `TrustServerSelfSignedCertificate` is enabled
- **Files**: `src/Paramore.Brighter.MessagingGateway.RMQ.Async/RmqTlsConfigurator.cs:60-61`, `src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqTlsConfigurator.cs:60-61`

**2. Race Condition in Sync Consumer**
- **Issue**: `RmqMessageConsumer.CreateConsumer()` was manually calling `HandleBasicConsumeOk()` after `BasicConsume()`, but RabbitMQ client already calls this method automatically as a callback
- **Impact**: Intermittent `InvalidOperationException: Operations that change non-concurrent collections must have exclusive access` due to concurrent modifications to non-thread-safe `HashSet` in `DefaultBasicConsumer`
- **Fix**: Removed the redundant manual call to `HandleBasicConsumeOk()` and added explanatory comment
- **Files**: `src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessageConsumer.cs:380-381`

## Testing

**All 32 mTLS tests (16 async + 16 sync) pass consistently.**

### Quick Test Commands (3 minutes total)

**Run all 32 mTLS tests (most efficient):**
```bash
dotnet test tests/Paramore.Brighter.RMQ.Async.Tests/Paramore.Brighter.RMQ.Async.Tests.csproj --filter "Category=MutualTLS" --framework net10.0 && \
dotnet test tests/Paramore.Brighter.RMQ.Sync.Tests/Paramore.Brighter.RMQ.Sync.Tests.csproj --filter "Category=MutualTLS" --framework net10.0
```

**OR run separately by category:**

```bash
# 1. Unit/configuration tests only (14 tests, no Docker required, ~5 seconds)
dotnet test tests/Paramore.Brighter.RMQ.Async.Tests/Paramore.Brighter.RMQ.Async.Tests.csproj \
  --filter "Category=MutualTLS&Requires!=Docker-mTLS" --framework net10.0

dotnet test tests/Paramore.Brighter.RMQ.Sync.Tests/Paramore.Brighter.RMQ.Sync.Tests.csproj \
  --filter "Category=MutualTLS&Requires!=Docker-mTLS" --framework net10.0

# 2. Acceptance/observability tests (18 tests, requires Docker, ~3 minutes)
# First time setup:
cd tests
./generate-test-certs.sh
docker-compose -f docker-compose.rabbitmq-mtls.yml up -d
cd ..

# Run Docker-dependent tests:
dotnet test tests/Paramore.Brighter.RMQ.Async.Tests/Paramore.Brighter.RMQ.Async.Tests.csproj \
  --filter "Requires=Docker-mTLS" --framework net10.0

dotnet test tests/Paramore.Brighter.RMQ.Sync.Tests/Paramore.Brighter.RMQ.Sync.Tests.csproj \
  --filter "Requires=Docker-mTLS" --framework net10.0
```

### Why These Commands Are Fast

- **Targets specific test projects** - Only builds RMQ-related projects (not the entire solution)
- **Single framework** - Uses `--framework net10.0` instead of testing on net8.0, net9.0, net10.0
- **Precise filtering** - Only runs the 32 mTLS tests, not all RMQ tests
- **3 minutes vs 3-5 hours** - Avoids building superfluous projects

### Test Coverage

Added comprehensive test coverage across four categories:

**Configuration Tests (14)**: Verify SSL plumbing, certificate loading edge cases, and validation logic
- Certificate object configuration
- File path configuration with passwords
- Invalid certificate handling
- Missing file error handling
- SSL option configuration

**Acceptance Tests (4)**: Verify actual mTLS connections against Docker RabbitMQ
- Basic publish with client certificate
- Round-trip publish and consume

**Observability Tests (10)**: Verify W3C Trace Context (TraceParent, TraceState, Baggage) and CloudEvents trace context survive mTLS connections
- TraceParent header preservation
- TraceState and Baggage preservation
- BrighterTracer.WriteProducerEvent integration
- CloudEvents trace context serialization

**Quorum Queue Tests (4)**: Verify mTLS works with RabbitMQ quorum queues
- Trace context with quorum queues
- Baggage with quorum queues

All tests include both async and sync variants to ensure cross-gateway uniformity.

### Test Stability

- Tests use `[Collection("RabbitMQ mTLS", DisableParallelization = true)]` to prevent TLS handshake race conditions
- Docker-dependent tests are marked with `[Trait("Requires", "Docker-mTLS")]` for selective execution
- All tests pass consistently on .NET 8.0, 9.0, and 10.0

## Infrastructure

- Added Docker Compose configuration for RabbitMQ with mTLS enabled (`tests/docker-compose.rabbitmq-mtls.yml`)
- Added RabbitMQ mTLS configuration file (`tests/rabbitmq-mtls.conf`)
- Added certificate generation script (`tests/generate-test-certs.sh`) for creating test CA, server, and client certificates
- Updated `.gitignore` to exclude generated test certificates

## Test Harness

Added a simple Web API sample in `samples/WebAPI/WebAPI_mTLS_TestHarness/` that demonstrates:
- Publishing and consuming messages over RabbitMQ with mTLS
- Both producer and consumer in a single application
- REST API with Swagger UI for easy testing

This provides a working example of the mTLS configuration and verifies end-to-end functionality.

## Usage Example

```csharp
// Option 1: Provide certificate object directly
var cert = new X509Certificate2("client-cert.pfx", "password");
var connection = new RmqMessagingGatewayConnection
{
    AmpqUri = new AmqpUriSpecification(new Uri("amqps://rabbitmq.example.com:5671")),
    Exchange = new Exchange("my.exchange"),
    ClientCertificate = cert,
    TrustServerSelfSignedCertificate = true // For dev/test only
};

// Option 2: Load from file path
var connection = new RmqMessagingGatewayConnection
{
    AmpqUri = new AmqpUriSpecification(new Uri("amqps://rabbitmq.example.com:5671")),
    Exchange = new Exchange("my.exchange"),
    ClientCertificatePath = "/path/to/client-cert.pfx",
    ClientCertificatePassword = "password",
    TrustServerSelfSignedCertificate = false // Production: validate properly
};

// Use with producer or consumer as normal
var producer = new RmqMessageProducer(connection);
await producer.SendAsync(message);
```

### Added
- `src/Paramore.Brighter.MessagingGateway.RMQ.Async/RmqTlsConfigurator.cs`
- `src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqTlsConfigurator.cs`
- `tests/docker-compose.rabbitmq-mtls.yml`
- `tests/rabbitmq-mtls.conf`
- `tests/generate-test-certs.sh`
- 6 test files with 32 comprehensive test cases
- `samples/WebAPI/WebAPI_mTLS_TestHarness/` (manual test harness)

### Modified
- `src/Paramore.Brighter.MessagingGateway.RMQ.Async/RmqMessageGateway.cs`
- `src/Paramore.Brighter.MessagingGateway.RMQ.Async/RmqMessagingGatewayConnection.cs`
- `src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessageGateway.cs`
- `src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessagingGatewayConnection.cs`
- `src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessageConsumer.cs`

## Notes

- Certificate configuration is fully backward compatible - existing code continues to work unchanged
- `TrustServerSelfSignedCertificate` defaults to `false` for production safety
- Certificate object takes precedence over file path if both are configured
- Tests use trait-based filtering rather than environment variables per feedback on ADR

## Security Considerations

- `TrustServerSelfSignedCertificate = true` should **only** be used in test/development environments
- In production, use properly signed certificates and validate the full certificate chain (`TrustServerSelfSignedCertificate = false`)
- Certificate passwords should be stored securely (e.g., Azure Key Vault, environment variables, not hardcoded)
- Certificates are loaded securely using the appropriate .NET APIs for the target framework

## Breaking Changes

None. This is a purely additive feature. Existing code without mTLS configuration continues to work as before.

## Known Issue

There is an intermittent test failure on .NET 10.0 only in CI:
- `RmqMessageProducerRequeuingMessageTests.When_posting_a_message_via_the_messaging_gateway`
- Passes consistently on .NET 9.0
- ASYNC version passes on both .NET 9.0 and 10.0
- SYNC now has identical logic to ASYNC

This appears to be a .NET 10.0-specific timing/environmental issue that requires further investigation. The core bug fix (missing `!` operator) is correct and working.

### Bug Fix: Corrected header filtering logic in sync RmqMessagePublisher

This PR includes a critical bug fix for message header handling that was discovered during mTLS testing.

**Issue:** In commit `76aeced5c` (September 2025), a refactoring of the sync `RmqMessagePublisher` accidentally inverted the header filtering logic in the `AddUserDefinedHeaders` method:

```csharp
// BEFORE (correct):
if (!_headersToReset.Any(htr => htr.Equals(header.Key)))
    headers.Add(header.Key, header.Value);

// AFTER (incorrect - missing NOT operator):
if (_headersToReset.Contains(header.Key))
{
    headers[header.Key] = header.Value;
}
```

**Impact:** This caused only Brighter's internal system headers (MESSAGE_TYPE, TOPIC, HANDLED_COUNT, DELIVERY_TAG, CORRELATION_ID) to be copied during message publishing and requeuing, while user-defined headers were discarded. This resulted in message body loss and data corruption during message requeuing operations.

**Fix:** Restored the correct logic by adding back the NOT operator:

```csharp
// CORRECTED:
if (!_headersToReset.Contains(header.Key))
{
    headers[header.Key] = header.Value;
}
```

This brings the sync implementation in line with the async version (`src/Paramore.Brighter.MessagingGateway.RMQ.Async/RmqMessagePublisher.cs:231`), which retained the correct logic throughout the refactoring.

**Discovery:** This bug was discovered during mTLS acceptance testing when messages were being requeued and custom headers were not preserved, causing test failures.

**Files Changed:**
- `src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessagePublisher.cs` (line 197)

---

### Note on Race Condition Fix

The race condition fix in `RmqMessageConsumer.CreateConsumer()` (removal of manual `HandleBasicConsumeOk()` call) has been reverted from this PR and will be addressed in a separate issue for better scope management. See RACE_CONDITION_ISSUE.md for details on creating the tracking issue.

---

### CODE REVIEW

**Comment by iancooper** [comment]
Likely back on Brighter stuff tomorrow

**Comment by darrenschwarz** [comment]
> Likely back on Brighter stuff tomorrow

HI @iancooper, ready for some eyes on :)

File: src/Paramore.Brighter.MessagingGateway.RMQ.Async/RmqTlsConfigurator.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
+#region Licence
+/* The MIT License (MIT)
+Copyright © 2014 Ian Cooper <ian_hammond_cooper@yahoo.co.uk>
```
Comment (iancooper) [inline src/Paramore.Brighter.MessagingGateway.RMQ.Async/RmqTlsConfigurator.cs]: This should really default to the author of the change. So in this case @darrenschwarz you. Our CLA leaves you the rights, but you grant us in perpetuity usage

File: src/Paramore.Brighter.MessagingGateway.RMQ.Async/RmqTlsConfigurator.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+using System;
+using System.IO;
+using System.Net.Security;
+using System.Security.Cryptography.X509Certificates;
+using RabbitMQ.Client;
+
+namespace Paramore.Brighter.MessagingGateway.RMQ.Async;
+
+/// <summary>
```
Comment (iancooper) [inline src/Paramore.Brighter.MessagingGateway.RMQ.Async/RmqTlsConfigurator.cs]: Not really a problem, but we only require /// comments on public types. Otherwise we would keep comments to "helps explain" and prefer obvious code.

File: src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqTlsConfigurator.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
+#region Licence
+/* The MIT License (MIT)
+Copyright © 2014 Ian Cooper <ian_hammond_cooper@yahoo.co.uk>
```
Comment (iancooper) [inline src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqTlsConfigurator.cs]: The comments here will be the same as in the async version

File: src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessageConsumer.cs [hunk 4]
```diff
@@ -377,7 +377,8 @@ private void CreateConsumer()
 
             Channel.BasicConsume(_queueName.Value, false, _consumerTag, false, false, SetQueueArguments(), _consumer);
 
-            _consumer.HandleBasicConsumeOk(_consumerTag);
+            // Note: HandleBasicConsumeOk is called automatically by the RabbitMQ client when BasicConsume completes
```
Comment (iancooper) [inline src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessageConsumer.cs]: This might be a good fix, but it feels out-of-scope to the change around mTLS and we should probably raise it under a separate issue

File: src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessagePublisher.cs [hunk 5]
```diff
@@ -194,7 +194,7 @@ private static void AddUserDefinedHeaders(Message message, Dictionary<string, ob
         {
             message.Header.Bag.Each(header =>
             {
-                if (_headersToReset.Contains(header.Key))
+                if (!_headersToReset.Contains(header.Key))
```
Comment (iancooper) [inline src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessagePublisher.cs:197]: Why did we reverse the sign here?

File: tests/Paramore.Brighter.RMQ.Async.Tests/MessagingGateway/When_configuring_mutual_tls_connection.cs [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+        }
+    }
+}
+
+// Acceptance tests verify actual mTLS connections to Docker RabbitMQ
+// These tests run against the actual transport configured in Docker (per ADR #3946)
+[Trait("Category", "RabbitMQ")]
+[Trait("Category", "MutualTLS")]
+[Trait("Requires", "Docker-mTLS")]
+public class RmqMutualTlsAcceptanceTests : IDisposable
```
Comment (iancooper) [inline tests/Paramore.Brighter.RMQ.Async.Tests/MessagingGateway/When_configuring_mutual_tls_connection.cs]: If we put the acceptance tests into a separate file, and within a separate folder within the test project, it is easier for someone working in an IDE to choose whether to run acceptance tests or just stick to developer tests while they iterate

One hint here is '[Trait("Requires", "Docker-mTLS")]` which indicates that you have different dependencies on this class than on the other class. A different category might also be a smell here

File: tests/Paramore.Brighter.RMQ.Async.Tests/MessagingGateway/When_configuring_mutual_tls_connection.cs [hunk 7]
```diff
@@ -0,0 +1,353 @@
+#region Licence
```
Comment (iancooper) [inline tests/Paramore.Brighter.RMQ.Async.Tests/MessagingGateway/When_configuring_mutual_tls_connection.cs]: We don't bother with the license file header in a test

File: tests/Paramore.Brighter.RMQ.Async.Tests/MessagingGateway/Acceptance/When_publishing_with_trace_context_over_mtls.cs [hunk 8]
```diff
@@ -0,0 +1,441 @@
+#region Licence
```
Comment (iancooper) [inline tests/Paramore.Brighter.RMQ.Async.Tests/MessagingGateway/Acceptance/When_publishing_with_trace_context_over_mtls.cs]: We don't bother with the license file header in a test

File: tests/Paramore.Brighter.RMQ.Async.Tests/MessagingGateway/When_publishing_with_trace_context_over_mtls.cs [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+            Exchange = new Exchange("test.mtls.exchange"),
+            ClientCertificatePath = _clientCertPath,
+            ClientCertificatePassword = CertPassword,
+            TrustServerSelfSignedCertificate = true
+        };
+
+        // Create message with trace context that will be serialized to CloudEvents format
+        var message = new Message(
+            new MessageHeader(
+                messageId: Guid.NewGuid().ToString(),
```
Comment (iancooper) [inline tests/Paramore.Brighter.RMQ.Async.Tests/MessagingGateway/When_publishing_with_trace_context_over_mtls.cs]: Picking this specific example, but applies generally to the PR. For an `Id` prefer `Id.Random()` over `Guild.NewGuid().ToString()`. This was a change with V10.

File: tests/Paramore.Brighter.RMQ.Sync.Tests/MessagingGateway/When_configuring_mutual_tls_connection.cs [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+    }
+}
+
+// Acceptance tests verify actual mTLS connections to Docker RabbitMQ
+// These tests run against the actual transport configured in Docker (per ADR #3946)
+[Trait("Category", "RabbitMQ")]
+[Trait("Category", "MutualTLS")]
+[Trait("Requires", "Docker-mTLS")]
+[Collection("RabbitMQ mTLS")]
+public class RmqMutualTlsAcceptanceTests : IDisposable
```
Comment (iancooper) [inline tests/Paramore.Brighter.RMQ.Sync.Tests/MessagingGateway/When_configuring_mutual_tls_connection.cs]: Same comment as with the Async version. Where we have different attributes, particularly `[Trait("Requires", "Docker-mTLS")]` we should use a different file and folder for these tests, to allow us to choose when to run them when working in the IDE, more easily.

File: tests/Paramore.Brighter.RMQ.Sync.Tests/MessagingGateway/When_configuring_mutual_tls_connection.cs [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+            Exchange = new Exchange("test.mtls.exchange.sync"),
+            ClientCertificatePath = _clientCertPath,
+            ClientCertificatePassword = CertPassword,
+            TrustServerSelfSignedCertificate = true  // Trust self-signed certificates in test environment
+        };
+
+        // Act
+        using var producer = new RmqMessageProducer(connection);
+        var message = new Message(
+            new MessageHeader(Guid.NewGuid().ToString(), "test.mtls.topic", MessageType.MT_EVENT),
```
Comment (iancooper) [inline tests/Paramore.Brighter.RMQ.Sync.Tests/MessagingGateway/When_configuring_mutual_tls_connection.cs]: The same comment as Sync and may appear multiple times. When working with an `Id` prefer to use `Id.Random()` over `Guid.NewGuid().ToString`

File: tests/generate-test-certs.sh [hunk 12]
```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
```
Comment (iancooper) [inline tests/generate-test-certs.sh:1]: We should have a Powershell file as well as some Brighter developers will be on Windows and prefer to use Powershell

**Comment by iancooper** [comment]
@darrenschwarz I am done commenting on this, for now, if you want to move it on.

File: src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessageConsumer.cs (hunk 4 above)
Comment (darrenschwarz) [reply src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessageConsumer.cs]: Agreed, will move to new issue.

File: src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessagePublisher.cs (hunk 5 above)
Comment (darrenschwarz) [reply src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessagePublisher.cs:197]: See explanation under "Bug Fix: Corrected header filtering logic in sync RmqMessagePublisher" in the description above.

**Comment by darrenschwarz** [comment]
> @darrenschwarz I am done commenting on this, for now, if you want to move it on.

All comments have been addressed, ready for review.

File: samples/WebAPI/WebAPI_mTLS_TestHarness/TodoApi/Properties/launchSettings.json [hunk 13]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (iancooper) [inline samples/WebAPI/WebAPI_mTLS_TestHarness/TodoApi/Properties/launchSettings.json:1]: Thanks for adding the samples, we appreciate it when new features come with examples of how to use them

File: src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessagePublisher.cs (hunk 5 above)
Comment (iancooper) [reply src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessagePublisher.cs:197]: Is this still reversing, though? The logic is that we only want to reset certain headers from a list. If they are in the list, we overwrite them. If we use the `!` operator, this means reset if the list does not contain the header.

**Comment by iancooper** [comment]
@darrenschwarz Even I am getting confused as to whether the sign reversal was fixed now

File: src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessagePublisher.cs (hunk 5 above)
Comment (darrenschwarz) [reply src/Paramore.Brighter.MessagingGateway.RMQ.Sync/RmqMessagePublisher.cs:197]: So the fix here is correct, unless I misunderstand the intention. This will ensure that we respect any user defined headers, and we are checking against the static system/blocked list of headers (_headersToReset).

**Comment by darrenschwarz** [comment]
> @darrenschwarz Even I am getting confused as to whether the sign reversal was fixed now

So the fix here is correct, unless I misunderstand the intention. This will ensure that we respect any user defined headers, and we are checking against the static system/blocked list of headers (_headersToReset). 

I'll run through with you when we catchup. If the fix is good, then I was thinking that we might want to do some minor renaming to make intention more apparent.

**Comment by iancooper** [comment]
> So the fix here is correct, unless I misunderstand the intention. This will ensure that we respect any user defined headers, and we are checking against the static system/blocked list of headers (_headersToReset).

I'll take a look. I need to refresh my memory of why we reset some headers.

Arguably the fix is out-of-scope for a change to mTLS. Now, human authors do this, because sometimes we stumble over a bug, and need to fix it to progress. I guess, when reviewing an agent's code, it's harder to see intent - why did you change this? - why did it help your code work.

**Comment by darrenschwarz** [comment]
> > So the fix here is correct, unless I misunderstand the intention. This will ensure that we respect any user defined headers, and we are checking against the static system/blocked list of headers (_headersToReset).
> 
> I'll take a look. I need to refresh my memory of why we reset some headers.
> 
> Arguably the fix is out-of-scope for a change to mTLS. Now, human authors do this, because sometimes we stumble over a bug, and need to fix it to progress. I guess, when reviewing an agent's code, it's harder to see intent - why did you change this? - why did it help your code work.

This was actually a human in the loop instance, I cam across it because tests were failing. :)

**Comment by darrenschwarz** [comment]
> > > So the fix here is correct, unless I misunderstand the intention. This will ensure that we respect any user defined headers, and we are checking against the static system/blocked list of headers (_headersToReset).
> > 
> > 
> > I'll take a look. I need to refresh my memory of why we reset some headers.
> > Arguably the fix is out-of-scope for a change to mTLS. Now, human authors do this, because sometimes we stumble over a bug, and need to fix it to progress. I guess, when reviewing an agent's code, it's harder to see intent - why did you change this? - why did it help your code work.
> 
> This was actually a human in the loop instance, I cam across it because tests were failing. :)

I'm happy to revert this and create another issue, and submit the fix as an isolated item.

**Comment by iancooper** [comment]
> I'm happy to revert this and create another issue, and submit the fix as an isolated item.

Nah, all good. Only gave the feedback in case it was an agent
